include/sleip/dynamic_array.hpp

```diff
@@ -33,250 +34,258 @@ struct dynamic_array
   using reverse_iterator       = std::reverse_iterator<iterator>;
   using const_reverse_iterator = std::reverse_iterator<const_iterator>;
 
+  static_assert(std::is_object_v<value_type>, "Only support object types");
+  static_assert(std::is_same_v<typename allocator_type::value_type, value_type>,
+                "Allocator's value type must match container's");
+
 private:
-  T*          data_ = nullptr;
-  std::size_t size_ = 0;
-  Allocator   alloc_;
+  struct deleter : boost::empty_value<Allocator, 0>
+  {
+    std::size_t size = 0;
+
+    deleter() noexcept(noexcept(Allocator()))
+      : boost::empty_value<Allocator, 0>(boost::empty_init_t{})
+    {
+    }
+
+    deleter(deleter const&) = delete;
+
+    deleter(deleter&& other) noexcept
+      : boost::empty_value<Allocator, 0>(
+          boost::empty_init_t{},
+          std::move(static_cast<boost::empty_value<Allocator, 0>&>(other).get()))
+      , size{other.size}
+    {
+      other.size = 0;
+    };
+
+    deleter(std::size_t size_)
+      : boost::empty_value<Allocator, 0>(boost::empty_init_t{}, Allocator())
+      , size{size_}
+    {
+    }
+
+    deleter(Allocator const& alloc, std::size_t size_)
+      : boost::empty_value<Allocator, 0>(boost::empty_init_t{}, alloc)
+      , size{size_}
+    {
+    }
+
+    auto
+    operator=(deleter const&) & -> deleter& = delete;
+
+    auto
+      operator=(deleter&& other) &
+      noexcept -> deleter&
+    {
+      boost::empty_value<Allocator, 0>::get() =
+        std::move(static_cast<boost::empty_value<Allocator, 0>&>(other).get());
+
+      size       = other.size;
+      other.size = 0;
+
+      return *this;
+    }
+
+    auto
+    get_allocator() const noexcept -> Allocator
```

Comment:
I would make this return by reference instead, and let the copy happen at the user interface edge.